Add MessagePack.Experimental package which includes SIMD(Single Instruction Multiple Data) accelerated primitive array formatters.

pCYSl5EDgo opened the pull request:
The latest LTS version of .NET Core is 3.1.

.NET Core 3.1 provides [Hardware Intrinsics](https://devblogs.microsoft.com/dotnet/hardware-intrinsics-in-net-core/).
The SIMD Intrinsics would make some formatters more faster. 

[Accelerating UTF-8 Decoding Using SIMD Instructions](https://researcher.watson.ibm.com/researcher/files/jp-INOUEHRS/IPSJPRO2008_SIMDdecoding.pdf)
Above article shows important implications for how to efficiently encode fixed-length elements to variable length.

AArnott commented:
Thanks for contributing, @pCYSl5EDgo. But I fail to see how this PR will improve performance at all. It just adds `HARDWARE_INTRINSICS_X86` as a preprocessor symbol but we don't have any source code that checks for this at the moment.
If you want to add such code with this PR, that's fine. But I don't think it makes sense as-is.

pCYSl5EDgo commented:
I will make some PR based on this PR.
This is the first and basic PR.
I add the symbol just for preparation.

AArnott commented:
Ok, I appreciate you wanting focused PRs, but I want at least some value in each PR. All this one does is add build time. 🙂
So please add the value to this one.
Also we'll need to make sure tests run on netcoreapp2.1 as well as netcoreapp3.1 to test all code paths.
Also if we really are adding x86-only paths to the code, we'll need to add automated testing on a non-x86 platform to make sure we don't regress those.

pCYSl5EDgo commented:
Thank you for your reminding me of the tests running on netcoreapp3.1!
I forgot it.

Ok, I'll make this PR involving one small improvement.
Thank you!

pCYSl5EDgo commented:
**Interim Report**

I tried to improve the performance of MessagePackWriter.Write(string) by examining that each char value of input string is in ASCII range.
It slows down because of iterating twice of input chars.

I'll try another type.

AArnott commented:
Thanks for measuring impact rather than just assuming the change makes things faster!

pCYSl5EDgo commented:
**Interim Report**

I add API `void MessagePackWriter.Write(sbyte[])` and `void MessagePackWriter.Write(ReadOnlySpan<sbyte>)`.

The table below compares the elapsed time between new Hardware Intrinsic code and MessagePack-CSharp v2.1.152.
In the test code, these 2 competitors encoded a 16MByte random sbyte array.

|          Method |      Mean |    Error |   StdDev |
|---------------- |----------:|---------:|---------:|
|   SerializeSimd |  41.64 ms | 0.810 ms | 1.188 ms |
| SerializeNoSimd | 119.85 ms | 2.391 ms | 2.559 ms |

pCYSl5EDgo commented:
**Interim Report**

I removed api `void MessagePackWriter.Write(sbyte[])`.
I changed the implementation of StandardClassFormatters.

I improved the performance of the serialization of `short[]`.

|              Method |      Mean |    Error |   StdDev |
|-------------------- |----------:|---------:|---------:|
|       SerializeSimd |  74.14 ms | 0.467 ms | 0.414 ms |
|     SerializeNoSimd | 243.51 ms | 1.664 ms | 1.475 ms |
|   SerializeSimdZero |  42.07 ms | 0.135 ms | 0.119 ms |
| SerializeNoSimdZero |  80.48 ms | 1.264 ms | 1.183 ms |

AArnott commented on src/MessagePack.UnityClient/Assets/Scripts/MessagePack/Formatters/StandardClassLibraryFormatter.cs:
```diff
@@ -35,6 +40,581 @@ public byte[] Deserialize(ref MessagePackReader reader, MessagePackSerializerOpt
+                        }
+                    }
+
+                    ReadOnlySpan<byte> shuffle = new byte[]
```
Is this array mutated or can we allocate it as a `static readonly` field? It's a huge array and we like to be as close to alloc-free as possible during serialization.

pCYSl5EDgo replied:
This is alloc-free code in Realease build.
I will move the `shuffle` table to another file( which is generated by T4).

pCYSl5EDgo replied:
[The sharplab.io sample code](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBLANgHwAEAmARgFgAoQgZgAIS6BhOgbyrs4fsJToFkAFAEo2HLhIBKMAIYATAPIA7XAE8AygAcZSgDzBVGGAD46qmQCsArlboBeOkpgB3OgaMBtALriJndpR+QXSkaL7BnKHhEejRwQCsYYERXFHJKXSx6SmkAJxJGZH5cUEAHGmFdAAMJXQAvgDctbUAZtCyYAAWdIIAbjJQjnTYSmaWNsK1AZWEeYJKwk3ZEnXhq5R1QA).

Since C#7.2(or 7.3 I don't remember...), the Roslyn compiler in Release Build avoids  the `byte[]` allocation when that `byte[]` is static and readonly.

AArnott replied:
Wow. That's very cool. Thanks for sharing.

AArnott replied:
However, I guess we need to know whether it's also alloc-free when Unity compiles it (in each of its many modes). It has an older C# compiler. We wouldn't want to regress that either.

pCYSl5EDgo replied:
Unity doesn't support .NET Core 3.1.
`HARDWARE_INTRINSICS_X86` is never defined.
So you don't have to mind that.

Unity2018.4 and later uses Roslyn compiler and enables this optimization.

AArnott commented on src/MessagePack.UnityClient/Assets/Scripts/MessagePack/Formatters/StandardClassLibraryFormatter.cs:
```diff
@@ -35,6 +40,581 @@ public byte[] Deserialize(ref MessagePackReader reader, MessagePackSerializerOpt
+
+#if HARDWARE_INTRINSICS_X86
+                const int Stride = 16;
+                if (Popcnt.IsSupported && inputLength >= Stride)
```
Does this `if` check make the code safe to run in all contexts that offer the API? IOW does `Popcnt.IsSupported` safely return `false` on all non-supporting processors?

pCYSl5EDgo replied:
[According to this official blog post](https://devblogs.microsoft.com/dotnet/hardware-intrinsics-in-net-core/#what-are-hardware-intrinsics), `Popcnt.IsSupported` property is trustable as far as .NET team implements this right.

AArnott commented on src/MessagePack.UnityClient/Assets/Scripts/MessagePack/Formatters/StandardClassLibraryFormatter.cs:
```diff
@@ -35,6 +40,581 @@ public byte[] Deserialize(ref MessagePackReader reader, MessagePackSerializerOpt
+                        0, 128, 1, 128, 2, 128, 3, 128, 4, 128, 5, 128, 6, 128, 7, 128,
+                        128, 0, 128, 1, 128, 2, 128, 3, 128, 4, 128, 5, 128, 6, 128, 7,
+                    };
+                    fixed (byte* pShuffle = shuffle)
```
I'm very impressed by what you wrote here. But I'm also very concerned about maintainability. I have no idea what any of this does and have never seen code like it before. The array above is a total mystery to me as to where it came from, and why "shuffling" is required baffles me.  If we keep this, IMO we would need a *lot* of code comments and links to docs or blogs that explain it.

Why in the world do we need to do all this when we could do a straight up memcpy of `sbyte[]` to the memory returned from `GetSpan`?

pCYSl5EDgo replied:
>Why in the world do we need to do all this when we could do a straight up memcpy of sbyte[] to the memory returned from GetSpan?

It's incorrect.
According to [current MessagePackWriter.Write(sbyte)](https://github.com/neuecc/MessagePack-CSharp/blob/68638214982c83d15ab0b0020445f9126ab017e3/src/MessagePack.UnityClient/Assets/Scripts/MessagePack/MessagePackWriter.cs#L227), Some sbyte values ranges from -33 to -128 have to be encoded and must not be just copied.

pCYSl5EDgo replied:
> I have no idea what any of this does and have never seen code like it before. 

It is known among the C/C++ programmers who use SIMD hardware intrinsics.
This is a good time for C#ers to learn and use SIMD programming.

> The array above is a total mystery to me as to where it came from, and why "shuffling" is required baffles me. 

I will explain where this `shuffle` table came from in the T4 file.

AArnott replied:
I look forward to the .tt file update.

> It's incorrect.
According to current MessagePackWriter.Write(sbyte), Some sbyte values ranges from -33 to -128 have to be encoded and must not be just copied.

Huh. Good point. I guess when we write out `byte[]` we skip the attempted compression of each byte because we precede it with a special msgpack 'binary' header, which we don't do for `sbyte[]`, so I guess we're on the hook to properly encode each one. But that's... really unfortunate and I would encourage folks who want good perf to simply cast their `sbyte[]` as a `byte[]` (using pointers) for faster and more compact encoding. But I guess I get what you're going for. Thanks for explaining.

pCYSl5EDgo replied:
> simply cast their sbyte[] as a byte[] (using pointers) for faster and more compact encoding. 
I did the same thing for personal usage...

pCYSl5EDgo replied:
References
- [Official Intel API Reference](https://software.intel.com/sites/landingpage/IntrinsicsGuide/)
- [The most useful website to understand the SIMD intruction behaviours](https://www.officedaytime.com/simd512e/)
- [.NET Hardware Intrinsics API](https://docs.microsoft.com/ja-jp/dotnet/api/system.runtime.intrinsics.x86?view=netcore-3.1)

pCYSl5EDgo replied:
The `sbyte[]` formatter is optimized differently than the other types...

The below table is measured by the `Int8ArrayBenchmarkMessagePackNoSimdVsMessagePackSimd` benchmark.

|              Method |      Mean |    Error |   StdDev |
|-------------------- |----------:|---------:|---------:|
|       SerializeSimd |  19.59 ms | 0.345 ms | 0.472 ms |
|     SerializeNoSimd | 116.50 ms | 1.306 ms | 1.221 ms |
|   SerializeSimdZero |  10.05 ms | 0.152 ms | 0.127 ms |
| SerializeNoSimdZero |  56.92 ms | 0.635 ms | 0.594 ms |
|    SerializeSimdM32 |  10.11 ms | 0.092 ms | 0.086 ms |
|  SerializeNoSimdM32 |  59.48 ms | 1.171 ms | 2.112 ms |
|    SerializeSimdM33 |  28.11 ms | 0.375 ms | 0.333 ms |
|  SerializeNoSimdM33 |  91.95 ms | 0.790 ms | 0.701 ms |

pCYSl5EDgo replied:
> But that's... really unfortunate and I would encourage folks who want good perf to simply cast their sbyte[] as a byte[] (using pointers) for faster and more compact encoding.

Actually I did same thing.
I change the formatter to cast `sbyte[]` as `byte[]` and encode it.
This is the benchmark result.

|              Method |      Mean |    Error |   StdDev |
|-------------------- |----------:|---------:|---------:|
|       SerializeSimd_ConvertByteArray |  11.05 ms | 0.206 ms | 0.426 ms |
|     SerializeNoSimd | 127.54 ms | 2.472 ms | 2.064 ms |
|   SerializeSimdZero_ConvertByteArray |  11.32 ms | 0.222 ms | 0.247 ms |
| SerializeNoSimdZero |  64.56 ms | 1.243 ms | 2.830 ms |
|    SerializeSimdM32_ConvertByteArray |  10.89 ms | 0.183 ms | 0.290 ms |
|  SerializeNoSimdM32 |  63.10 ms | 1.178 ms | 1.157 ms |
|    SerializeSimdM33_ConvertByteArray |  10.77 ms | 0.150 ms | 0.133 ms |
|  SerializeNoSimdM33 |  97.38 ms | 0.998 ms | 0.885 ms |

My PR is not that bad compared to that which converts to byte[].

AArnott commented on src/MessagePack.UnityClient/Assets/Scripts/MessagePack/Formatters/StandardClassLibraryFormatter.cs:
```diff
@@ -35,6 +40,581 @@ public byte[] Deserialize(ref MessagePackReader reader, MessagePackSerializerOpt
         }
     }
 
+    public sealed class SByteArrayFormatter : IMessagePackFormatter<sbyte[]>
```
I'm curious why use `sbyte` here? Who uses `sbyte[]`? If there's a perf improvement to be made in writing out an `sbyte[]`  array, wouldn't that also apply to `byte[]`, which is far more popular?

pCYSl5EDgo replied:
[The specialized Formatter](https://github.com/neuecc/MessagePack-CSharp/blob/master/src/MessagePack.UnityClient/Assets/Scripts/MessagePack/Formatters/StandardClassLibraryFormatter.cs#L19) has already existed for `byte[]`. I think that formatter is fastest. I don't need to improve that.

Implemented : `sbyte[]`, `short[]`
Work In Progress : `int[]`, `ushort[]`, `uint[]`

The reason of `sbyte[]` is that its implementation difficulty seemed easy to me compared to others.
I get used to implementing improved formatters.

I will write `int[]` formatter which seems more difficult to implement than `sbyte[]`.

AArnott replied:
well, before you go to any more work, I'd like to feel settled on what you've done so far so you don't waste effort if we're not going to take the PR ultimately anyway. I'm not saying we won't... I'm just saying that since you submitted some, let us review this and understand it enough to justify your continued effort here.

AArnott commented on src/MessagePack.UnityClient/Assets/Scripts/MessagePack/MessagePackWriter.cs:
```diff
@@ -1210,7 +1214,7 @@ private static void WriteBigEndian(ulong value, Span<byte> span)
             // solves heuristic length check
 
             // ensure buffer by MaxByteCount(faster than GetByteCount)
```
this comment is now stale if you remove the use of `GetMaxByteCount`.

pCYSl5EDgo replied:
Thank you!

AArnott commented on src/MessagePack.UnityClient/Assets/Scripts/MessagePack/MessagePackWriter.cs:
```diff
@@ -1210,7 +1214,7 @@ private static void WriteBigEndian(ulong value, Span<byte> span)
 
             // ensure buffer by MaxByteCount(faster than GetByteCount)
-            bufferSize = StringEncoding.UTF8.GetMaxByteCount(characterLength) + 5;
+            bufferSize = (characterLength * 3) + 5;
```
Where did you get this `3` multiplier? UTF8's longest encoded character is 4 [according to this](https://stackoverflow.com/questions/9533258/what-is-the-maximum-number-of-bytes-for-a-utf-8-encoded-character).  
As it is, the `GetMaxByteCount` method looks relatively simple so I don't see how this is a significant perf win. A general rule for accuracy vs. performance is "[as simple as possible *but no simpler*.](https://championingscience.com/2019/03/15/everything-should-be-made-as-simple-as-possible-but-no-simpler/#:~:text=%E2%80%9CEverything%20should%20be%20made%20as%20simple%20as%20possible%2C,a%20great%20statement%20of%20how%20to%20conduct%20science.)" Based on the [.NET implementation of this method](https://source.dot.net/#System.Private.CoreLib/UTF8Encoding.cs,199682d3d0340ec9) it looks like you're over-simplifying this, potentially at the risk of under-estimating the maximum required buffer size.

pCYSl5EDgo replied:
Within the BMP, UTF8 requires a maximum of three bytes; UTF16 requires one character.
Outside the BMP range, UTF16 uses two characters; UTF8 always requires four bytes.
That's two bytes UTF8 per UTF16 character.

So it's estimated to be up to 3 bytes UTF8 per one character UTF16.

pCYSl5EDgo replied:
.NET implementation is too complex because they want to handle the invalid UTF16 strings.
It is undefined how to encode the invalid UTF16 to valid UTF8 bytes.

AArnott replied:
What is BMP? And can you point to an RFC (or another doc of similar notoriety) to back up what you're saying? I really don't want to regress this and the first RFC I found said 4 was the magic number.

pCYSl5EDgo replied:
> Also: did you actually see GetMaxByteCount taking too long in a perf trace? 
No, I didn't so I agree with you.

pCYSl5EDgo commented:
**Interim Report**

`bool[]` Serialize and Deserailize is optimized.
This optimization has resulted in a speed increase of 2 to 20 times.

```ini
BenchmarkDotNet=v0.12.1, OS=Windows 10.0.18363.997 (1909/November2018Update/19H2)
Intel Core i7-8750H CPU 2.20GHz (Coffee Lake), 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=3.1.400
  [Host]   : .NET Core 3.1.6 (CoreCLR 4.700.20.26901, CoreFX 4.700.20.31603), X64 RyuJIT
  ShortRun : .NET Core 3.1.6 (CoreCLR 4.700.20.26901, CoreFX 4.700.20.31603), X64 RyuJIT

Job=ShortRun  IterationCount=3  LaunchCount=1  
WarmupCount=3  

```
|               Method |     Size |             Mean |            Error |          StdDev |
|--------------------- |--------- |-----------------:|-----------------:|----------------:|
|        **SerializeSimd** |       **64** |         **147.2 ns** |         **78.32 ns** |         **4.29 ns** |
|      SerializeNoSimd |       64 |         336.8 ns |        251.45 ns |        13.78 ns |
|      DeSerializeSimd |       64 |         235.1 ns |         33.54 ns |         1.84 ns |
|    DeserializeNoSimd |       64 |         317.9 ns |         36.10 ns |         1.98 ns |
|   SerializeSimdFalse |       64 |         151.3 ns |         37.09 ns |         2.03 ns |
| SerializeNoSimdFalse |       64 |         329.2 ns |        291.10 ns |        15.96 ns |
|    SerializeSimdTrue |       64 |         150.0 ns |         44.78 ns |         2.45 ns |
|  SerializeNoSimdTrue |       64 |         346.4 ns |        141.81 ns |         7.77 ns |
|        **SerializeSimd** |     **1024** |         **323.8 ns** |      **1,545.91 ns** |        **84.74 ns** |
|      SerializeNoSimd |     1024 |       6,351.3 ns |      2,354.72 ns |       129.07 ns |
|      DeSerializeSimd |     1024 |         308.4 ns |         46.42 ns |         2.54 ns |
|    DeserializeNoSimd |     1024 |       4,895.7 ns |      1,940.78 ns |       106.38 ns |
|   SerializeSimdFalse |     1024 |         260.3 ns |        185.66 ns |        10.18 ns |
| SerializeNoSimdFalse |     1024 |       3,833.0 ns |        392.59 ns |        21.52 ns |
|    SerializeSimdTrue |     1024 |         238.3 ns |        109.91 ns |         6.02 ns |
|  SerializeNoSimdTrue |     1024 |       3,846.9 ns |        577.95 ns |        31.68 ns |
|        **SerializeSimd** | **16777216** |  **11,380,270.8 ns** |  **3,346,765.39 ns** |   **183,447.50 ns** |
|      SerializeNoSimd | 16777216 | 109,296,553.3 ns | 39,192,596.66 ns | 2,148,278.47 ns |
|      DeSerializeSimd | 16777216 |   5,376,443.0 ns |  2,713,570.75 ns |   148,739.97 ns |
|    DeserializeNoSimd | 16777216 | 113,424,266.7 ns | 92,964,361.48 ns | 5,095,690.35 ns |
|   SerializeSimdFalse | 16777216 |  12,428,593.2 ns | 15,857,118.12 ns |   869,182.15 ns |
| SerializeNoSimdFalse | 16777216 |  62,635,550.0 ns | 39,142,827.42 ns | 2,145,550.45 ns |
|    SerializeSimdTrue | 16777216 |  11,856,341.1 ns | 17,099,923.28 ns |   937,304.50 ns |
|  SerializeNoSimdTrue | 16777216 |  64,057,362.5 ns | 60,948,291.25 ns | 3,340,781.50 ns |

AArnott commented:
@neuecc How do you feel about the change?

The metrics that @pCYSl5EDgo shares look incredible, and the PR includes some nice documentation in the .tt files to help folks understand where to learn more about the code. 

That said, I remain concerned that:

1. This PR raises the level of coding from a 200 to 400 level in the impacted areas. I still haven't taken time to understand it myself, and I fear doing so may be a significant time investment.
1. The perf benefits may be focused on fairly narrow scenarios such as long primitive arrays that don't impact enough people to be worth the maintenance cost.

My concerns aren't 'blocking', but I share anyway so @neuecc considers them along with benefits to make a call on this.

neuecc commented:
The target is only a primitive array, right?
Also, it doesn't apply to byte[], so I think it's niche.

How about creating a new Experimental package (MessagePack.Experimental) and putting it together as SimdArrayResolver there?

pCYSl5EDgo commented:
Targets are `sbyte[]`, `short[]`, `int[]`, `bool[]`, and `float[]`.

> MessagePack.Experimental

Okay, I start working with this proposal.

pCYSl5EDgo commented:
@AArnott @neuecc 

I made a `MessagePack.Experimental` package which provides `MessagePack.Experimental.Resolvers.PrimitiveArrayResolver`.
`MessagePack.Experimental.csproj` targets .Net Core 3.1.
Now, `MessagePack.csproj` doesn't need netcoreapp3_1 and `HARDWARE_INTRINSICS_X86` preprocessor symbol.

pCYSl5EDgo commented:
Add double[] formatter.

``` ini

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.18363.1049 (1909/November2018Update/19H2)
Intel Core i7-8750H CPU 2.20GHz (Coffee Lake), 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=3.1.401
  [Host]  : .NET Core 3.1.7 (CoreCLR 4.700.20.36602, CoreFX 4.700.20.37001), X64 RyuJIT
  LongRun : .NET Core 3.1.7 (CoreCLR 4.700.20.36602, CoreFX 4.700.20.37001), X64 RyuJIT

Job=LongRun  IterationCount=100  LaunchCount=3  
WarmupCount=15  

```
|                                   Method |     Size |             Mean |           Error |           StdDev |           Median |
|----------------------------------------- |--------- |-----------------:|----------------:|-----------------:|-----------------:|
|   **SerializeSingleInstructionMultipleData** |       **64** |         **227.3 ns** |         **3.22 ns** |         **16.52 ns** |         **224.0 ns** |
| SerializeNoSingleInstructionMultipleData |       64 |         604.4 ns |         5.13 ns |         25.96 ns |         596.5 ns |
|   **SerializeSingleInstructionMultipleData** |     **1024** |       **1,490.5 ns** |        **18.70 ns** |         **93.75 ns** |       **1,463.4 ns** |
| SerializeNoSingleInstructionMultipleData |     1024 |       9,230.7 ns |       354.94 ns |      1,827.70 ns |       8,519.3 ns |
|   **SerializeSingleInstructionMultipleData** | **16777216** | **142,284,585.5 ns** | **2,387,494.64 ns** | **12,208,105.14 ns** | **137,715,500.0 ns** |
| SerializeNoSingleInstructionMultipleData | 16777216 | 267,093,020.7 ns | 4,500,563.82 ns | 23,053,582.08 ns | 259,778,500.0 ns |

AArnott commented:
I like the idea of the Experimental package. Should we close this PR then? We could perhaps mention the Experimental package on our README so it's more discoverable for folks who have similar goals.

AArnott commented:
I'm re-reviewing, but so far almost all the diffs I see in the existing projects are style changes. Please revert *everything* that's unrelated to your perf work. You can submit another PR with the style changes if you feel so inclined and we can weigh those separately.

AArnott commented:
Never mind, I'll take care of it while I'm reviewing.

AArnott commented on src/MessagePack.Experimental/PrimitiveArrayResolver.cs:
```diff
@@ -0,0 +1,31 @@
+
+using MessagePack.Formatters;
+
+namespace MessagePack.Experimental.Resolvers
```
If we use the actual namespace and type names that we would in the primary library, we can eventually *move* these types from Experimental to the main library without a binary breaking change in the future. Any concerns with dropping `Experimental` from the namespace?

pCYSl5EDgo replied:
I do not have any concerns.
I had forgotten the perspective "binary breaking change in the future".
Thank you for your dropping `Experimental`!

AArnott reviewed (approved):
Looks good. Thanks for contributing!

AArnott commented:
But IMO we should target 2.2 with this change, which is only days away from being released anyway.

pCYSl5EDgo commented:
Thank you for your review and approval!

>MessagePack_2_1_165.dll with the one from the nuget package by that version. The one placed here previously was slightly different and I don't know why, but using the official build seems prudent.

I'm worried the benchmark project will not work.
I made the change because I was referring to the similar benchmark project "benchmark/SerializerBenchmark".
"benchmark/SerializerBenchmark" has "MessagePack_1_7_3_6.dll".
"MessagePack_1_7_3_6.dll" is different from Nuget's one.

The differences between "MessagePack_1_7_3_6.dll" and Nuget Official one

- File name
- Assembly name
- Module name

[C#'s extern alias official explanation](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/extern-alias) saids that `extern alias` needs different assembly name.
I made a change to the Nuget one to change its assembly and module name.

pCYSl5EDgo commented:
CI seems to fail in the step of `dotnet restore`.
There seems to be a problem about the version dependency in `MessagePack.Generator.Tests`.
CI saids that `MessagePack.Generator.Tests` depends on v3.4.0 and `MessagePack.GeneratorCore` depends on v3.6.0.
I read both csproj and do not find 3.6.0 dependency...

I do not know how to fix the CI failure. :(

AArnott commented:
I'll take a look at the CI break.

Regarding assembly names, it should not be necessary to change the assembly name if the assembly version is already unique, but I will confirm this.

AArnott commented:
The CI break resolved itself. I suspect you had retargeted to the `v2.2` branch before I merged `master` into `v2.2` yesterday, so the build didn't include the fix I included in #1036 for the package restore failure. That also would explain why you didn't see the failure locally: your branch hadn't actually merged with v2.2 either. Anyway, although this PR currently shows a build failure, I rebuilt the PR in Azure Pipelines and it succeeded so if we push to your PR again, it'll work.

Regarding the assembly name, yes the assembly names must be unique, and yours wasn't because you were building in a v2.1 branch and had checked in messagepack with an assembly version of 2.1.0.0, which matched what was built in that branch. So once your change merges with the v2.2 branch and the assembly version changes to 2.2.0.0, it works. So I'm going to push a change to your PR that merges with v2.2 then reverts the custom build of messagepack.dll so it's just the standard one since at that point we won't need the assembly name change.

AArnott commented:
@neuecc I'm satisfied with this. Are you?

neuecc reviewed (approved):
good.

pCYSl5EDgo commented:
> CI break

I appreciate your fix.

> Assembly Name

Oh, I'll study it again. Thank you for your explanatation.

AArnott commented:
@pCYSl5EDgo do you mind if I squash your PR instead of merge it, given it has 47 commits?

pCYSl5EDgo commented:
@AArnott 
I do not mind it. Thank you.

AArnott commented:
Thanks for your contribution, @pCYSl5EDgo !
